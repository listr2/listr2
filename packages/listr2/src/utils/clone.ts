import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
