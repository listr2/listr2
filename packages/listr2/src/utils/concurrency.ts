// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
