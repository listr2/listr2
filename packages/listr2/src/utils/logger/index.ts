export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
