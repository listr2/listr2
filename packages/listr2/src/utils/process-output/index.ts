export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
