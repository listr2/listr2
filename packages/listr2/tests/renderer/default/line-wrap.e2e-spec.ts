import { Listr } from '@root'
import type { MockProcessOutput } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput } from '@tests/utils'

describe('default renderer: option: line wrap', () => {
  const message =
    'THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE'

  const messageWithNewLines =
    'THIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE'

  const messageWithEmptyLines =
    'THIS IS A LONG LONG MESSAGE\n\n\nAGE\nTHIS ISESSAGE\nTHIS IG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE\nTHIS IS A LONG LONG MESSAGE'

  process.stdout.isTTY = true
  process.stdout.columns = 80

  const output: MockProcessOutput = {} as MockProcessOutput

  process.stdout.isTTY = true

  beforeEach(async() => {
    mockProcessOutput(output)
  })

  afterEach(async() => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  it.each([
    ['truncate', message],
    ['wrap', message],
    ['truncate', messageWithNewLines],
    ['wrap', messageWithNewLines]
  ])('should %s long strings', async(format, m) => {
    let err: Error

    try {
      await new Listr(
        [
          {
            title: `This task will with formating the output with ${format} output.`,
            task: async(_, task): Promise<void> => {
              task.output = m
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true, formatOutput: format as any }
        }
      ).run()
    } catch (e: any) {
      err = e
    }

    expect(err).toBeFalsy()
    expectProcessOutputToMatchSnapshot(output, '31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA')
  })

  it.each([
    ['truncate', message],
    ['wrap', message],
    ['truncate', messageWithNewLines],
    ['wrap', messageWithNewLines]
  ])('should %s long titles', async(format, m) => {
    let err: Error

    try {
      await new Listr(
        [
          {
            title: m,
            task: async(_, task): Promise<void> => {
              task.output = `This task will with formating the title with ${format} output.`
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true, formatOutput: format as any }
        }
      ).run()
    } catch (e: any) {
      err = e
    }

    expect(err).toBeFalsy()
    expectProcessOutputToMatchSnapshot(output, 'effxCyY3v0UwCvij5FA2kc2XllnFtU4B')
  })

  it.each([
    [true, 'truncate', messageWithEmptyLines],
    [false, 'truncate', messageWithEmptyLines]
  ])('should skip empty lines %s', async(rel, format, m) => {
    let err: Error

    try {
      await new Listr(
        [
          {
            title: `This will skip empty lines on output: ${rel}`,
            task: async(_, task): Promise<void> => {
              task.output = m
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: {
            lazy: true,
            formatOutput: format as any,
            removeEmptyLines: rel
          }
        }
      ).run()
    } catch (e: any) {
      err = e
    }

    expect(err).toBeFalsy()
    expectProcessOutputToMatchSnapshot(output, 'lEAjB2T3MtNUO3ufIOFMvwZ6eLpeIfhx')
  })
})
