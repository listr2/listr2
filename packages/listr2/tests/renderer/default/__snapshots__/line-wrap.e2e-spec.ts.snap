// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default renderer: option: line wrap should skip empty lines false: lEAjB2T3MtNUO3ufIOFMvwZ6eLpeIfhx-exit 1`] = `[]`;

exports[`default renderer: option: line wrap should skip empty lines false: lEAjB2T3MtNUO3ufIOFMvwZ6eLpeIfhx-stderr 1`] = `[]`;

exports[`default renderer: option: line wrap should skip empty lines false: lEAjB2T3MtNUO3ufIOFMvwZ6eLpeIfhx-stdout 1`] = `
[
  "[?25l",
  "‚óº This will skip empty lines on output: false
",
  "[2K[1A[2K[G‚ùØ This will skip empty lines on output: false
",
  "[2K[1A[2K[G‚ùØ This will skip empty lines on output: false
  ‚Ä∫ THIS IS A LONG LONG MESSAGE


    AGE
    THIS ISESSAGE
    THIS IG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This will skip empty lines on output: false
",
  "[2K[1A[2K[G",
  "‚úî This will skip empty lines on output: false
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should skip empty lines true: lEAjB2T3MtNUO3ufIOFMvwZ6eLpeIfhx-exit 1`] = `[]`;

exports[`default renderer: option: line wrap should skip empty lines true: lEAjB2T3MtNUO3ufIOFMvwZ6eLpeIfhx-stderr 1`] = `[]`;

exports[`default renderer: option: line wrap should skip empty lines true: lEAjB2T3MtNUO3ufIOFMvwZ6eLpeIfhx-stdout 1`] = `
[
  "[?25l",
  "‚óº This will skip empty lines on output: true
",
  "[2K[1A[2K[G‚ùØ This will skip empty lines on output: true
",
  "[2K[1A[2K[G‚ùØ This will skip empty lines on output: true
  ‚Ä∫ THIS IS A LONG LONG MESSAGE
    AGE
    THIS ISESSAGE
    THIS IG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This will skip empty lines on output: true
",
  "[2K[1A[2K[G",
  "‚úî This will skip empty lines on output: true
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should truncate long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-exit 1`] = `[]`;

exports[`default renderer: option: line wrap should truncate long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-exit 2`] = `[]`;

exports[`default renderer: option: line wrap should truncate long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-stderr 1`] = `[]`;

exports[`default renderer: option: line wrap should truncate long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-stderr 2`] = `[]`;

exports[`default renderer: option: line wrap should truncate long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will with formating the output with truncate output.
",
  "[2K[1A[2K[G‚ùØ This task will with formating the output with truncate output.
",
  "[2K[1A[2K[G‚ùØ This task will with formating the output with truncate output.
  ‚Ä∫ THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LO‚Ä¶
",
  "[2K[1A[2K[1A[2K[G‚úî This task will with formating the output with truncate output.
",
  "[2K[1A[2K[G",
  "‚úî This task will with formating the output with truncate output.
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should truncate long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-stdout 2`] = `
[
  "[?25l",
  "‚óº This task will with formating the output with truncate output.
",
  "[2K[1A[2K[G‚ùØ This task will with formating the output with truncate output.
",
  "[2K[1A[2K[G‚ùØ This task will with formating the output with truncate output.
  ‚Ä∫ THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will with formating the output with truncate output.
",
  "[2K[1A[2K[G",
  "‚úî This task will with formating the output with truncate output.
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should truncate long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-exit 1`] = `[]`;

exports[`default renderer: option: line wrap should truncate long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-exit 2`] = `[]`;

exports[`default renderer: option: line wrap should truncate long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-stderr 1`] = `[]`;

exports[`default renderer: option: line wrap should truncate long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-stderr 2`] = `[]`;

exports[`default renderer: option: line wrap should truncate long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-stdout 1`] = `
[
  "[?25l",
  "‚óº THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG‚Ä¶
",
  "[2K[1A[2K[G‚ùØ THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG‚Ä¶
",
  "[2K[1A[2K[G‚ùØ THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG‚Ä¶
  ‚Ä∫ This task will with formating the title with truncate output.
",
  "[2K[1A[2K[1A[2K[G‚úî THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG‚Ä¶
",
  "[2K[1A[2K[G",
  "‚úî THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG‚Ä¶
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should truncate long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-stdout 2`] = `
[
  "[?25l",
  "‚óº THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  ‚Ä∫ This task will with formating the title with truncate output.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G",
  "‚úî THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should wrap long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-exit 1`] = `[]`;

exports[`default renderer: option: line wrap should wrap long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-exit 2`] = `[]`;

exports[`default renderer: option: line wrap should wrap long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-stderr 1`] = `[]`;

exports[`default renderer: option: line wrap should wrap long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-stderr 2`] = `[]`;

exports[`default renderer: option: line wrap should wrap long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will with formating the output with wrap output.
",
  "[2K[1A[2K[G‚ùØ This task will with formating the output with wrap output.
",
  "[2K[1A[2K[G‚ùØ This task will with formating the output with wrap output.
  ‚Ä∫ THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG
    LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS
     A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE
     THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
     MESSAGE THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will with formating the output with wrap output.
",
  "[2K[1A[2K[G",
  "‚úî This task will with formating the output with wrap output.
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should wrap long strings: 31Xbkp3zzsfDqlG3p8uAIotvBXNr92hA-stdout 2`] = `
[
  "[?25l",
  "‚óº This task will with formating the output with wrap output.
",
  "[2K[1A[2K[G‚ùØ This task will with formating the output with wrap output.
",
  "[2K[1A[2K[G‚ùØ This task will with formating the output with wrap output.
  ‚Ä∫ THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
    THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will with formating the output with wrap output.
",
  "[2K[1A[2K[G",
  "‚úî This task will with formating the output with wrap output.
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should wrap long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-exit 1`] = `[]`;

exports[`default renderer: option: line wrap should wrap long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-exit 2`] = `[]`;

exports[`default renderer: option: line wrap should wrap long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-stderr 1`] = `[]`;

exports[`default renderer: option: line wrap should wrap long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-stderr 2`] = `[]`;

exports[`default renderer: option: line wrap should wrap long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-stdout 1`] = `
[
  "[?25l",
  "‚óº THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
   MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A
  LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS
   IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
  MESSAGE THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
  MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG
   LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS
  A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
  MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG
   LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS
  A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  ‚Ä∫ This task will with formating the title with wrap output.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
   MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A
  LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS
   IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
  MESSAGE THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G",
  "‚úî THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
   MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A
  LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS
   IS A LONG LONG MESSAGE THIS IS A LONG LONG MESSAGE THIS IS A LONG LONG
  MESSAGE THIS IS A LONG LONG MESSAGE
",
  "[?25h",
]
`;

exports[`default renderer: option: line wrap should wrap long titles: effxCyY3v0UwCvij5FA2kc2XllnFtU4B-stdout 2`] = `
[
  "[?25l",
  "‚óº THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚ùØ THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  ‚Ä∫ This task will with formating the title with wrap output.
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G‚úî THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G",
  "‚úî THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
  THIS IS A LONG LONG MESSAGE
",
  "[?25h",
]
`;
